sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuApproximatePercentile.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import com.nvidia.spark.rapids.GpuCast.{recursiveDoColumnar}
+
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentile
+import org.apache.spark.sql.catalyst.util.ArrayData
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.{CudfTDigest, GpuAggregateFunction}
+import org.apache.spark.sql.types.{ArrayType, DataType, DataTypes}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * The ApproximatePercentile function returns the approximate percentile(s) of a column at the given
+ * percentage(s). A percentile is a watermark value below which a given percentage of the column
+ * values fall. For example, the percentile of column `col` at percentage 50% is the median of
+ * column `col`.
+ *
+ * This function supports partial aggregation.
+ *
+ * The GPU implementation uses t-digest to perform the initial aggregation (see
+ * `updateExpressions` / `mergeExpressions`) and then applies the ApproxPercentileFromTDigestExpr`
+ * expression to compute percentiles from the final t-digest (see `evaluateExpression`).
+ *
+ * There are two different data types involved here. The t-digests are a map of centroids
+ * (`Map[mean: Double -> weight: Double]`) represented as `List[Struct[Double, Double]]` and
+ * the final output is either a single double or an array of doubles, depending on whether
+ * the `percentageExpression` parameter is a single value or an array.
+ *
+ * @param child child expression that can produce column value with `child.eval()`
+ * @param percentageExpression Expression that represents a single percentage value or
+ *                             an array of percentage values. Each percentage value must be between
+ *                             0.0 and 1.0.
+ * @param accuracyExpression Integer literal expression of approximation accuracy. Higher value
+ *                           yields better accuracy, the default value is
+ *                           DEFAULT_PERCENTILE_ACCURACY.
+ */
+case class GpuApproximatePercentile (
```

Comment:
I see that most other aggregate classes are put into `AggregateFunctions.scala`. Should we put this class into it to? In the future, we may organize the aggregates into a folder by a similar way as in Spark.

Reply:
This is large enough that I think it can stand on its own. I am fine if we want to put the aggregations into a separate aggregations package like Spark does. We have not done a great job with breaking up a single large rapids package.